pilot/pkg/bootstrap/server.go

```diff
@@ -1281,3 +1287,60 @@ func (s *Server) initStatusManager(_ *PilotArgs) {
 		return nil
 	})
 }
+
+// initLocalSecretListener creates a new secret listener to listen for
+// secrets which represent new multi cluster participating clusters in a mesh.
+func (s *Server) initLocalSecretListener(args *PilotArgs) {
+	// Check if the istiod's kube client API endpoint is the same as the local cluster API endpoint
+	// If they are the same, then the Istiod is in cluster Istiod, otherwise, it will be an
+	// external istiod running outside of the config cluster, the localKubeClient and the
+	// istiod kube client endpoint should have been different.
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Errorf("Could not get istiod incluster configuration: %v", err)
+		return
+	}
+	log.Info("Successfully retrieved incluster config.")
+
+	localKubeClient, err := kubelib.NewClient(kubelib.NewClientConfigForRestConfig(config))
+	if err != nil {
+		log.Errorf("Could not create a client to access local cluster API server: %v", err)
+		return
+	}
+	log.Infof("Successfully created in cluster kubeclient at %s", localKubeClient.RESTConfig().Host)
+	if s.kubeClient.RESTConfig().Host == localKubeClient.RESTConfig().Host {
```

Comment:
This check makes me feel a little uneasy...

I thought we already have an explicitly flag for "external control plane"?

Reply:
@howardjohn thanks so much for reviewing this so quickly. I talked to Frank about this, the external control plane flag alone is not good enough. I will talk to Frank again why we cannot solely rely on that flag. Would like to know if you are ok we are also watching the namespace that istiod is running in so that the secrets do not have to be created in the config cluster to join a new cluster to the mesh? Thanks John.

Reply:
The `ExternalIstiod` flag says `If this is set to true, one Istiod will control remote clusters including CA`, which doesn't necessarily imply an official "external control plane". It could also be a primary cluster that is controlling an istiodless remote in a Primary-Remote multicluster deployment. [This PR](https://github.com/istio/istio/pull/38283) is actually trying to make `ExternalIstiod` the default, so it will almost always be true.

You need to check if the config is in a different cluster, so that you can add a second secret listener for the local cluster in that case.

Reply:
That check does exactly that. @frankbu. Not sure if there is any other way for doing this.

Reply:
The question is why does @howardjohn feel uneasy about your kubeclient check? If that's an issue, I suggested a possible alternative way (last sentence in my previous comment).

Reply:
`Host` (and probably most fields or combination of fields) is not a unique identifier for a cluster. You can have many clusters with the same host, and many hosts for the same cluster. It seems very unreliable.